pkg/service/group_service_factory.go

```diff
@@ -16,69 +16,82 @@ type NsGroupService struct {
-	subscriptionService               *SubscriptionService
-	tokenService                      *TokenService
-	trustedOauthIssuerService         *TrustedOauthIssuerService
-	userGroupService                  *UserGroupService
+	applicationServerService          *ApplicationServerService
 	userPolicyService                 *UserPolicyService
+	userGroupService                  *UserGroupService
+	subnetService                     *SubnetService
+	controllerService                 *ControllerService
+	fibreChannelSessionService        *FibreChannelSessionService
 	userService                       *UserService
-	versionService                    *VersionService
-	volumeCollectionService           *VolumeCollectionService
-	volumeService                     sdkprovider.VolumeService
+	clientCredentialService           *ClientCredentialService
+	protectionScheduleService         *ProtectionScheduleService
+	initiatorGroupService             *InitiatorGroupService
+	accessControlRecordService        *AccessControlRecordService
+	activeDirectoryMembershipService  *ActiveDirectoryMembershipService
+	fibreChannelPortService           *FibreChannelPortService
+	protocolEndpointService           *ProtocolEndpointService
 	witnessService                    *WitnessService
+	jobService                        *JobService
+	auditLogService                   *AuditLogService
+	poolService                       sdkprovider.PoolService
+	volumeCollectionService           *VolumeCollectionService
+	subscriberService                 *SubscriberService
+	diskService                       *DiskService
+	fibreChannelInitiatorAliasService *FibreChannelInitiatorAliasService
+	groupService                      *GroupService
+	trustedOauthIssuerService         *TrustedOauthIssuerService
+	softwareVersionService            *SoftwareVersionService
+	ldapDomainService                 *LdapDomainService
+	networkConfigService              *NetworkConfigService
 }
 
 // NewNsGroupService - initializes NsGroupService
-func NewNsGroupService(ip, username, password, apiVersion string, synchronous bool) (gs *NsGroupService, err error) {
+// This function is depreciated. Call NewNimbleGroupService() to initialize a group service
+func NewNsGroupService(ip, username, password, apiVersion string, synchronous bool, clientOpts ...client.ClientOption) (gs *NsGroupService, err error) {
 	if apiVersion != "v1" {
 		return nil, fmt.Errorf("NewNsGroupService: unsupported %s sdk API version", apiVersion)
 	}
-	client, err := client.NewClient(ip, username, password, apiVersion, synchronous)
+	client, err := client.NewClient(ip, username, password, apiVersion, synchronous, clientOpts...)
 	if err != nil {
 		return nil, err
 	}
 	return &NsGroupService{ip: ip, client: client}, nil
 }
 
+func NewNimbleGroupService(clientOpts ...client.ClientOption) (gs *NsGroupService, err error) {
+	groupMgmtClient, err := client.NewClient("", "", "", "", false, clientOpts...)
```

Comment:
hmm I'm not sure what you mean. Here's what I'm getting;
1. `NewNsGroupService() (depreciated)` calls on `NewClient()` so `NewClient` has to support positional arguments. 
2. The parameters of NewNimbleGroupService() is of type `ClientOption`, which returns a function that updates `GroupMgmtClient`. If NewClient() only accepts positional arguments, we have to find a way to extract the parameters from ClientOption, convert them to string or boolean, then pass them to `NewClient()`. That doesn't sound like a good solution.

I think what you're suggesting is to create another function in `service` that does what `NewClient()` does-- but receives optional arguments instead of positional arguments? Is that correct?

Reply:
Imagine for a moment that the only change we make in group_mgmt_client.go is to add another positional argument to NewClient().  All the functional options stuff can be handled here in group_service_factory.go.  I think I threw you off initially by putting my comment at the NewClient() code point when it should have been here.

With that premise, we could define a groupServiceOptions structure and a ServiceOption type that acts on it.  The NewNimbleGroupService() function can process the options into a groupServiceOptions object (lower case because it is not part of the external interface).  It can then call NewClient() with positional arguments that are extracted from the groupServiceOptions object.  With this approach, the default option values would be defined here in group_service_factory.go, which makes sense because they should be documented here at the interface an SDK consumer will use.

It might seem like we are creating duplication by defining a struct that is similar to GroupMgmtClient.  However, it is not really the same.  Notice how you needed to add Host and ApiVersion fields to GroupMgmtClient just to support the options processing.  I think the focus should be on defining an extensible interface that is clear and easy to use at the consumer entry point (here).  Subsequently perhaps we could find ways to streamline the interactions between the service package and the client package, but there's no rush there since the client interfaces are more easily changed.